lib/mini_racer/truffleruby.rb

```diff
@@ -0,0 +1,305 @@
+
+        backtrace = e.backtrace.map { |line| line.sub('(eval)', '(mini_racer)') }
+        raise error_class, e.message, backtrace
+      end
+
+      def convert_js_to_ruby(value)
+        case value
+        when true, false, Integer, Float
+          value
+        else
+          if value.nil?
+            nil
+          elsif value.respond_to?(:call)
+            MiniRacer::JavaScriptFunction.new
+          elsif value.respond_to?(:to_str)
+            value.to_str
+          elsif value.respond_to?(:to_ary)
+            value.to_ary.map do |e|
+              if e.respond_to?(:call)
+                nil
+              else
+                convert_js_to_ruby(e)
+              end
+            end
+          elsif is_time(value)
+            js_date_to_time(value)
+          elsif is_symbol(value)
+            js_symbol_to_symbol(value)
+          else
+            object = value
+            h = {}
+            object.instance_variables.each do |member|
+              v = object[member]
+              unless v.respond_to?(:call)
+                h[member.to_s] = convert_js_to_ruby(v)
+              end
+            end
+            h
+          end
+        end
+      end
+
+      def is_time(value)
+        f = eval_in_context("(x) => { return x instanceof Date };")
+        f.call(value)
+      end
+
+      def js_date_to_time(value)
+        f = eval_in_context("(x) => { return x.getTime(x) };")
+        millis = f.call(value)
+        Time.at(Rational(millis, 1000))
+      end
+
+      def is_symbol(value)
+        f = eval_in_context("(x) => { return typeof x === 'symbol' };")
+        f.call(value)
+      end
+
+      def js_symbol_to_symbol(value)
+        f = eval_in_context("(x) => { var r = x.description; return r === undefined ? 'undefined' : r };")
+        f.call(value).to_sym
+      end
+
+      def convert_ruby_to_js(value)
+        case value
+        when nil, true, false, Integer, Float, String
+          value
+        when Array
+          value.map { |e| convert_ruby_to_js(e) }
+        when Hash
+          h = @js_object.new
+          value.each_pair do |k,v|
+            h[convert_ruby_to_js(k)] = convert_ruby_to_js(v)
+          end
+          h
+        when Symbol
+          value.to_s
+        when Time
+          eval_in_context("new Date(#{value.to_f * 1000})")
+        when DateTime
+          eval_in_context("new Date(#{value.to_time.to_f * 1000})")
+        else
+          "Undefined Conversion"
+        end
+      end
+
+    def encode(string)
+      raise ArgumentError unless string
+      string.encode(::Encoding::UTF_8)
+    end
+
+    class_eval <<-'RUBY', "(mini_racer)", 1
+        def eval_in_context(code); @context.eval('js', code); end
+    RUBY
+
+  end
+
+  class Isolate
+    def init_with_snapshot(snapshot)
+        # TruffleRuby does not have a 1-1 concept with isolate.
+        # However, isolate can hold a napshot, and code and ASTs are shared between contexts.
+        @snapshot = snapshot
```

Comment:
I kept using isolate instead of a dummy since it seems we can support it's function to hold/reuse a snapshot.